File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/workspace/InitGating.kt
package com.tyhp.lang.workspace

const val INIT_ACTION = "init"
const val INIT_DONT_ASK_AGAIN_KEY = "tyhp.init.dontAskAgain"

data class InitPromptContext(
    val isTyhpFile: Boolean,
    val hasOwner: Boolean,
    val hasAncestorTyhpJson: Boolean,
    val hasForcedProject: Boolean,
    val hasContentRoot: Boolean,
    val dontAskAgain: Boolean,
    val promptedThisSession: Boolean,
)

/**
 * True when a Tyhp editor is open, the file is not owned by any project,
 * there is no ancestor `tyhp.json` up to the content root, and the user has
 * not already dismissed the prompt. `tyhp.projectPath` suppresses the prompt.
 */
fun shouldPromptInit(context: InitPromptContext): Boolean {
    return context.isTyhpFile &&
        !context.hasOwner &&
        !context.hasAncestorTyhpJson &&
        !context.hasForcedProject &&
        context.hasContentRoot &&
        !context.dontAskAgain &&
        !context.promptedThisSession
}

fun buildInitArgs(): List<String> = listOf(INIT_ACTION, "--yes")

fun initErrorMessage(stderr: String?, stdout: String?, exitCode: Int): String {
    val err = stderr?.trim().orEmpty()
    if (err.isNotEmpty()) {
        return err
    }
    val out = stdout?.trim().orEmpty()
    if (out.isNotEmpty()) {
        return out
    }
    return "exit code $exitCode"
}

File: tyhp-lang/vscode/src/workspace/initGating.ts
/**
 * Pure helpers for “Tyhp: Initialize Project”.
 *
 * CLI (`DisplayHelp.InitHelp`): `tyhp init [--yes / -y] [directory]`.
 * `--yes` accepts template defaults without prompting. Do not pass
 * `--tyhp-project` — that flag requires an existing file
 * (`CliStartup.TryValidateProjectFile`).
 *
 * Prompt only when THIS file has no include-owner and no ancestor `tyhp.json`
 * (walk up to the workspace root). An ancestor that does not include the file
 * stays silent (TextMate only). `tyhp.projectPath` suppresses the prompt.
 */

export const INIT_ACTION = "init";
export const INIT_DONT_ASK_AGAIN_KEY = "tyhp.init.dontAskAgain";

export interface InitPromptContext {
    languageId: string;
    hasOwner: boolean;
    hasAncestorTyhpJson: boolean;
    hasForcedProject: boolean;
    hasWorkspaceFolder: boolean;
    dontAskAgain: boolean;
    promptedThisSession: boolean;
}

/**
 * True when a Tyhp editor is open, the file is not owned by any project,
 * there is no ancestor `tyhp.json` up to the workspace root, and the user
 * has not already dismissed the prompt.
 */
export function shouldPromptInit(context: InitPromptContext): boolean {
    return (
        context.languageId === "tyhp" &&
        !context.hasOwner &&
        !context.hasAncestorTyhpJson &&
        !context.hasForcedProject &&
        context.hasWorkspaceFolder &&
        !context.dontAskAgain &&
        !context.promptedThisSession
    );
}

/**
 * Argv for a non-interactive `tyhp init` (not including the executable).
 * Cwd must be the workspace folder that should receive `tyhp.json`.
 */
export function buildInitArgs(): string[] {
    return [INIT_ACTION, "--yes"];
}
